src/decoder/predictor.rs

```diff
@@ -0,0 +1,156 @@
+use crate::bytecast;
+
+pub trait Wrapping {
+    fn wrapping_add(&self, other: Self) -> Self;
+}
+
+impl Wrapping for u8 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        u8::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for u16 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        u16::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for u32 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        u32::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for u64 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        u64::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for i8 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        i8::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for i16 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        i16::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for i32 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        i32::wrapping_add(*self, other)
+    }
+}
+
+impl Wrapping for i64 {
+    fn wrapping_add(&self, other: Self) -> Self {
+        i64::wrapping_add(*self, other)
+    }
+}
+
+pub fn rev_hpredict_nsamp<T>(
+    image: &mut [T],
+    (width, height): (usize, usize), // Size of the block
+    img_width: usize, // Width of the decoding result (this distinction is needed for tiles)
+    samples: usize,
+) where
+    T: Copy + Wrapping,
+{
+    for row in 0..height {
+        for col in samples..width * samples {
+            let prev_pixel = image[(row * img_width * samples + col - samples)];
+            let pixel = &mut image[(row * img_width * samples + col)];
+            *pixel = pixel.wrapping_add(prev_pixel);
+        }
+    }
+}
+
+pub fn fp_predict<T>(
```

Comment:
I'm not sure it makes sense for `fp_predict` to be generic. If there was just a `fp_predict_f32` and `fp_predict_f64` they could both be much simpler. I think you should be able to do something along the lines of: 
```rust
fn fp_predict_f32(buf: &mut [f32], (width, height): (usize, usize), res_width: usize, samples: usize) {
    for row in 0..height {
        rev_hpredict_nsamp(AsNeMutBytes::as_ne_mut_bytes(
            &mut buf[row * res_width * simples_per_pixel..][..width * samples_per_pixel]), ...);
        for i in 0..(width*samples_per_pixel) {
             buf[i] = f32::from_bits(buf[i].to_bits().from_be())
        }
    }
}

Reply:
FWIW there are 16- and 24-bit floating point TIFFs, the 16-bit version not being that rare in the wild so we might have an interest in supporting it at some point. Libtiff supports both.